WindSpigot-Server/src/main/java/ga/windpvp/windspigot/async/entitytracker/ThreadSafeTrackerEntry.java

```diff
@@ -0,0 +1,205 @@
+package ga.windpvp.windspigot.async.entitytracker;
+
+import java.util.List;
+import net.minecraft.server.Entity;
+import net.minecraft.server.EntityHuman;
+import net.minecraft.server.EntityPlayer;
+import net.minecraft.server.EntityTracker;
+import net.minecraft.server.EntityTrackerEntry;
+import net.minecraft.server.Packet;
+
+/*
+ * This is an entry for entity trackers that is thread safe. All public methods accessed by  
+ * multiple threads are synchronized if not already synchronized.
+ */
+public class ThreadSafeTrackerEntry extends EntityTrackerEntry {
+
+	public ThreadSafeTrackerEntry(EntityTracker entityTracker, Entity entity, int b, int c, boolean flag) {
+		super(entityTracker, entity, b, c, flag);
+	}
+
+	public ThreadSafeTrackerEntry(Entity entity, int b, int c, boolean flag) {
+		super(entity, b, c, flag);
+	}
+
+	private boolean synchronize() {
+		return !Thread.holdsLock(this);
+	}
+
+	@Override
+	public boolean equals(Object object) {
+		if (synchronize()) {
+			synchronized (this) {
+				return super.equals(object);
+			}
+		} else {
+			return super.equals(object);
+		}
+	}
+
+	@Override
+	public int hashCode() {
+		if (synchronize()) {
+			synchronized (this) {
+				return super.hashCode();
+			}
+		} else {
+			return super.hashCode();
```

Comment:
This code is reused a lot, it should be moved to a method such as `synchronizeIfRequired(Object, Runnable)` (I believe Runnable would be the correct Interface to use, correct me if I’m wrong)

Reply:
Wouldn't this result in a creating a bunch of new Runnables every method call?

Reply:
The runnables can be cached, also I believe the compiler already optimizes this